File: AriBodaghee/BODAGHEE.TEST.PARITCLE/TEST.PARITCLE/Lab 4 - Dart/source/Emitter.cpp
#include "Emitter.h"
#include "CGame.h"

Emitter::Emitter()
{
	Data = new FlyWeight();
	reAnimate = false;
}

Emitter::~Emitter()
{
	for(unsigned int i = 0; i < Particle_List.size(); i++)
		delete (Particle*)(Particle_List[i]);

	delete Data;

	Particle_List.clear();
}

//		Trilogy of EVIL!:
//			Copy constructor
Emitter::Emitter(const Emitter& E)
{
	for(unsigned int i = 0; i < E.Particle_List.size(); i++)
		Particle_List.push_back(E.Particle_List[i]);

	Data = E.Data;

	
}

//			Assignment operator
Emitter& Emitter::operator=(const Emitter& E)
{
	if(this == &E)
		return *this;
	else
		*this = E;

	return *this;
}


void Emitter::Init()
{
	Particle* tempParticle;

	for (int i = 0; i < Data->MaxParticles; i++)
	{
		tempParticle = new Particle();
		tempParticle->Life = -1;
		Particle_List.push_back(tempParticle);

		if(Continuous == false)
			ReSpwn(tempParticle);
	}
}

void Emitter::Update( float DT )
{
	for (unsigned int i = 0; i < Particle_List.size(); i++)
	{
		if (Particle_List[i]->Life > 0)
		{
			if (Data->FadeEffect == 2)
				Particle_List[i]->A -= Data->FadeValue * DT;
			else if (Data->FadeEffect == 1)
				Particle_List[i]->A += Data->FadeValue * DT;

			if (Data->RotationEffect == 2)
				Particle_List[i]->Rotation -= Data->RotationValue / 10.0f * DT;
			else if (Data->RotationEffect == 1)
				Particle_List[i]->Rotation += Data->RotationValue / 10.0f * DT;

			if (Data->ScaleEffect == 2)
				Particle_List[i]->Size_p += (Data->ScaleValue / 100.0f) * DT;
			else if (Data->ScaleEffect == 1)
				Particle_List[i]->Size_p -= (Data->ScaleValue / 100.0f) * DT;

			if (Data->RedEffect == 2)
				Particle_List[i]->R += Data->RedValue * DT;
			else if (Data->RedEffect == 1)
				Particle_List[i]->R -= Data->RedValue * DT;

			if (Data->GreenEffect == 2)
				Particle_List[i]->G += Data->GreenValue * DT;
			else if (Data->GreenEffect == 1)
				Particle_List[i]->G -= Data->GreenValue * DT;

			if (Data->BlueEffect == 2)
				Particle_List[i]->B += Data->BlueValue * DT;
			else if (Data->BlueEffect == 1)
				Particle_List[i]->B -= Data->BlueValue * DT;

			Particle_List[i]->Update(DT);
		}
		else if(Continuous)
		{
			ReSpwn(Particle_List[i]);
		}
		else if( reAnimate )
		{
			ReSpwn(Particle_List[i]);

			if ( i >= Particle_List.size() - 1 )
			{
				reAnimate = false;
			}
		}

				Data->EmitterPosX += Data->EmitterVelX * DT;
				Data->EmitterPosY += Data->EmitterVelY * DT;
	}

}

void Emitter::ReSpwn(Particle* P)
{
	P->Life = (float)(rand() % (int)(Data->MaxLife * 1000));
	P->Size_p = Data->fSize;

	if (Data->RotationEffect > 0)
                P->Rotation = (float)(rand() % 100 / 11);
            else
                P->Rotation = 0.0f;

	if(rand() % 1000 >= 500)
		P->VelX = rand() % (int)(Data->MaxVelX - Data->MinVelX + 1) + Data->MinVelX;
	else
		P->VelX = -(rand() % (int)(Data->MaxVelX - Data->MinVelX + 1) + Data->MinVelX);

	if(rand() % 1000 <= 500)
		P->VelY = rand() % (int)(Data->MaxVelY - Data->MinVelY + 1) + Data->MinVelY;
	else
		P->VelY = -(rand() % (int)(Data->MaxVelY - Data->MinVelY + 1) + Data->MinVelY);


	if (Data->FadeEffect == 0 || Data->FadeEffect == 2)
		P->A = 255;
	else
		P->A = 0;

	if (Data->RedEffect == 0 || Data->RedEffect == 1)
		P->R = 255;
	else
		P->R = 0;

	if (Data->GreenEffect == 0 || Data->GreenEffect == 1)
		P->G = 255;
	else
		P->G = 0;

	if (Data->BlueEffect == 0 || Data->BlueEffect == 1)
		P->B = 255;
	else
		P->B = 0;

	if (Data->SpwnPattern == 1)
	{
		if(rand() % 1000 <= 500)
			P->PosX = Data->EmitterPosX - rand() % ((int)Data->MaxSpwnRadius + 1);
		else
			P->PosX = Data->EmitterPosX + rand() % ((int)Data->MaxSpwnRadius + 1);

		if(rand() % 1000 <= 500)
			P->PosY = Data->EmitterPosY - rand() % ((int)Data->MaxSpwnRadius + 1);
		else
			P->PosY = Data->EmitterPosY + rand() % ((int)Data->MaxSpwnRadius + 1);
	}

	if (Data->SpwnPattern == 3)
	{
		P->PosX = Data->EmitterPosX - (rand() % (int)(Data->SpwnAreaWidth * 2 + 1) - Data->SpwnAreaWidth) / 2;
		P->PosY = Data->EmitterPosY - (rand() % (int)(Data->SpwnAreaHeight * 2 + 1) - Data->SpwnAreaHeight) / 2;
	}

	if (Data->SpwnPattern == 2)
	{
		P->PosX = (float)(rand() % CGame::GetInstance()->GetScreenW());
		P->PosY = (float)(rand() % CGame::GetInstance()->GetScreenH());
	}
	
}

void Emitter::Render()
{
	DWORD tempBlends[2] = {0};
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->GetRenderState(D3DRS_SRCBLEND, &tempBlends[0]);
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->GetRenderState(D3DRS_DESTBLEND, &tempBlends[1]);

		CSGD_Direct3D::GetInstance()->GetSprite()->Flush();
		int image;
		Emitter* tempE = this;

		if (tempE->Image == -1)
			return;
		else
			image = tempE->Image;

		if (tempE->Data->UseBlend)
		{
			CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_SRCBLEND, tempE->Data->Source);
			CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_DESTBLEND, tempE->Data->Dest);
		}

		for (int i = tempE->Particle_List.size() - 1; i >= 0; i--)
		{
			Particle* tempP = tempE->Particle_List[i];

			if (tempP->Life <= 0)
				continue;

			int OffSetX = (int)(tempP->PosX - (float)CSGD_TextureManager::GetInstance()->GetTextureWidth(image) * tempP->Size_p / 2.0f);
			int OffsetY = (int)(tempP->PosY - (float)CSGD_TextureManager::GetInstance()->GetTextureHeight(image) * tempP->Size_p / 2.0f);

			if(tempP->Rotation == 0)
				CSGD_TextureManager::GetInstance()->Draw(image, OffSetX, OffsetY, tempP->Size_p, tempP->Size_p,NULL,0,0,0,D3DCOLOR_ARGB((int)tempP->A,(int)tempP->R,(int)tempP->G,(int)tempP->B)); 
			else
				CSGD_TextureManager::GetInstance()->Draw(image, (int)tempP->PosX, (int)tempP->PosY, tempP->Size_p, tempP->Size_p,NULL,0,0,0,D3DCOLOR_ARGB((int)tempP->A,(int)tempP->R,(int)tempP->G,(int)tempP->B));         

		}
		CSGD_Direct3D::GetInstance()->GetSprite()->Flush();


	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_SRCBLEND, tempBlends[0]);
	CSGD_Direct3D::GetInstance()->GetDirect3DDevice()->SetRenderState(D3DRS_DESTBLEND, tempBlends[1]);
	CSGD_Direct3D::GetInstance()->GetSprite()->Flush();
}

File: Ari/BodagheeParticleTool/BodagheeParticleTool/Emitter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using BodagheeParticleTool;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;

namespace ParticleTool
{
    class Emitter
    {
        public List<Particle> Particle_List = new List<Particle>();
        public bool continuous;
        public bool reAnimate;
        public int image;
        public string imageFile;

        Color colorStart;
        Color colorEnd;
        private float fSize;
        private float maxLife;
        private float minVelX;
        private float maxVelX;
        private float minVelY;
        private float maxVelY;
        private float spwnAreaWidth;
        private float spwnAreaHeight;
        private float scaleStart;
        private float scaleEnd;
        private float rotation;
        private int spwnPattern;
        private int source;
        private int dest;
        private float emitterPosX;
        private float emitterPosY;
        private int maxParticles;
        private float emitterVelX;
        private float emitterVelY;

        public Color ColorStart { get { return colorStart; } set { colorStart = value; } }
        public Color ColorEnd { get { return colorEnd; } set { colorEnd = value; } }
        public float FSize { get { return fSize; } set { fSize = value; } }
        public float MaxLife { get { return maxLife; } set { maxLife = value; } }
        public float MinVelX { get { return minVelX; } set { minVelX = value; } }
        public float MaxVelX {get { return maxVelX; } set { maxVelX = value; } }
        public float MinVelY { get { return minVelY; } set { minVelY = value; } }
        public float MaxVelY { get { return maxVelY; } set { maxVelY = value; } }
        public float SpwnAreaWidth {get { return spwnAreaWidth; }set { spwnAreaWidth = value; }}
        public float SpwnAreaHeight{ get { return spwnAreaHeight; }set { spwnAreaHeight = value; }}
        public float ScaleStart{ get { return scaleStart; } set { scaleStart = value; } }
        public float ScaleEnd{ get { return scaleEnd; }set { scaleEnd = value; }}
        public float Rotation{ get { return rotation; } set { rotation = value; } }
        public int SpwnPattern {get { return spwnPattern; } set { spwnPattern = value; } }
        public int Source{get { return source; }set { source = value; } }
        public int Dest{get { return dest; } set { dest = value; } }
        public float EmitterPosX{get { return emitterPosX; }set { emitterPosX = value; }}
        public float EmitterPosY{ get { return emitterPosY; }set { emitterPosY = value; }}
        public int MaxParticles { get { return maxParticles; }set { maxParticles = value; } }
        public float EmitterVelX {get { return emitterVelX; } set { emitterVelX = value; } }
        public float EmitterVelY {get { return emitterVelY; }set { emitterVelY = value; }}

        public Emitter()
        {
            image = 0;
            colorStart = Color.White;
            colorEnd = Color.White;
            fSize = 1.0f;
            maxLife = 1.0f;
            minVelX = 0;
            maxVelX = 0;
            minVelY = 0;
            maxVelY = 0;
            spwnAreaWidth = 2;
            spwnAreaHeight = 2;
            scaleStart = 0.0f;
            scaleEnd = 0.0f;
            rotation = 0.0f;
            spwnPattern = 0;
            source = 2;
            dest = 3;
            emitterPosX = 0;
            emitterPosY = 0;
            maxParticles = 0;
            emitterVelX = 0;
            emitterVelY = 0;
            continuous = true;
            reAnimate = false;
        }

        public void Update(float deltaTime)
        {
            for (int i = 0; i < Particle_List.Count(); i++)
            {
                Particle_List[i].rotation += rotation * deltaTime;
                Particle_List[i].life -= 1000 * deltaTime;
                Particle_List[i].Color = ColorOperator.Lerp(ColorEnd, ColorStart, Particle_List[i].life / 1000);
                
                Particle_List[i].posX += Particle_List[i].velX * deltaTime;
                Particle_List[i].posY += Particle_List[i].velY * deltaTime;

                if (scaleStart <= scaleEnd )
                {
                    Particle_List[i].size_p += 0.5f * deltaTime;
                    if (Particle_List[i].size_p >= scaleEnd)
                    {
                        Particle_List[i].size_p += 0;
                    }
                }
                if (scaleStart >= scaleEnd )
                {
                    Particle_List[i].size_p -= 0.5f * deltaTime;
                    if (Particle_List[i].size_p <= scaleStart)
                    {
                        Particle_List[i].size_p += 0;
                    }
                }

                EmitterPosX += EmitterVelX * deltaTime;
                EmitterPosY += EmitterVelY * deltaTime;

                if (Particle_List[i].Life < 0 && continuous)
                {
                    Respawn(Particle_List[i]);
                }
                //else if (reAnimate)
                //{
                // Respawn(Particle_List[i]);

                // if (i >= Particle_List.Count() - 1)
                // {
                //     reAnimate = false;
                // }
                // }
            }
        }
        public void Respawn(Particle p)
        {
            Random rand = new Random();
            p.Life = (float)(rand.Next(0, (int)MaxLife * 1000));
            p.Size_p = FSize;

            p.VelX = rand.Next((int)(MinVelX), (int)(MaxVelX));
            p.VelY = rand.Next((int)(MinVelY), (int)(MaxVelY));

            p.PosX = EmitterPosX - (rand.Next(0, (int)SpwnAreaWidth * 2 + 1) - SpwnAreaWidth) / 2;
            p.PosY = EmitterPosY - (rand.Next(0, (int)SpwnAreaHeight * 2 + 1) - SpwnAreaHeight) / 2;
        }
        public void Render()
        {
            int tempBlend1 = 0;
            int tempBlend2 = 0;

            tempBlend1 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.SourceBlend;
            tempBlend2 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.DestinationBlend;

            SGP.ManagedDirect3D.Instance.Sprite.Flush();

            SGP.ManagedDirect3D.Instance.Device.SetRenderState(Microsoft.DirectX.Direct3D.RenderStates.SourceBlend, this.Source);
            SGP.ManagedDirect3D.Instance.Device.SetRenderState(Microsoft.DirectX.Direct3D.RenderStates.DestinationBlend, this.Dest);

            for (int i = 0; i < this.Particle_List.Count; i++)
            {
                if (this.Particle_List[i].Life > 0)
                {
                    int OffSetX = (int)((int)(SGP.ManagedTextureManager.Instance.GetTextureWidth(image) >> 1) * this.Particle_List[i].Size_p);
                    int OffSetY = (int)((int)(SGP.ManagedTextureManager.Instance.GetTextureHeight(image) >> 1) * this.Particle_List[i].Size_p);

                    SGP.ManagedTextureManager.Instance.Draw(image, (int)this.Particle_List[i].PosX,
                        (int)this.Particle_List[i].PosY, this.Particle_List[i].Size_p, this.Particle_List[i].Size_p, Rectangle.Empty, OffSetX, OffSetY,
                        this.Particle_List[i].rotation, this.Particle_List[i].Color.ToArgb());
                }
            }
            SGP.ManagedDirect3D.Instance.Sprite.Flush();
            tempBlend1 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.SourceBlend;
            tempBlend2 = (int)SGP.ManagedDirect3D.Instance.Device.RenderState.DestinationBlend;
        }
    }
}
